ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,256 @@
-    kafkaStreams.add(queryMetadata.getKafkaStreams());
+  public void init(List<KafkaMetric> list) {
+
   }
 
+
   @Override
-  public void onDeregister(final QueryMetadata query) {
-    final KafkaStreams streams = query.getKafkaStreams();
-    kafkaStreams.remove(streams);
+  public void configure(Map<String, ?> map) {
+    String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    synchronized (lock) {
+      if (registeredNodeMetrics != null) {
+        return;
+      }
+      registeredNodeMetrics = new ArrayList<>();
+
+      final MetricName nodeAvailable = metricRegistry.metricName("node-storage-available", "ksql-utilization-metrics");
+      final MetricName nodeTotal = metricRegistry.metricName("node-storage-total", "ksql-utilization-metrics");
+      final MetricName nodeUsed = metricRegistry.metricName("node-storage-used", "ksql-utilization-metrics");
+      registeredNodeMetrics.add(nodeAvailable);
+      registeredNodeMetrics.add(nodeTotal);
+      registeredNodeMetrics.add(nodeUsed);
+
+      metricRegistry.addMetric(
+        nodeAvailable,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+      metricRegistry.addMetric(
+        nodeTotal,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+      metricRegistry.addMetric(
+        nodeUsed,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+    }
   }
 
   @Override
-  public void run() {
-    logger.info("Reporting Observability Metrics");
-    final Long currentTime = time.milliseconds();
-    // here is where we would report metrics
-    lastSampleTime = currentTime;
+  public void metricChange(final KafkaMetric metric) {
+    if (!metric.metricName().name().equals("total-sst-files-size")) {
+      return;
+    }
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final String queryIdTag = metric.metricName().tags().getOrDefault("thread-id", "");
+    Pattern pattern = Pattern.compile("(?<=query_)(.*?)(?=-)");
+    Matcher matcher = pattern.matcher(queryIdTag);
+    final String queryId = matcher.find() ? matcher.group(1) : "";
+    // if we haven't seen a task for this query yet
+    if (!metricsSeen.containsKey(queryId)) {
+      metricRegistry.addMetric(
+        metricRegistry.metricName(
+          "query-storage-usage",
+          "ksql-utilization-metrics",
+          ImmutableMap.of("query-id", queryId)),
+        (Gauge<BigInteger>) (config, now) -> computeQueryMetric(queryId)
+      );
+      metricsSeen.put(queryId, new ArrayList<>());
+    }
+    final TaskStorageMetric newMetric = new TaskStorageMetric(
```

Comment:
I think you can simplify this bit with a few changes:
- make `metricsSeen` a map from query id to a map from task id to `TaskStorageMetric` (e.g. `Map<String, Map<String, TaskStorageMetric>`)
- there's no need for `resetAndUpdateMetric` - you just need to call `add` on `TaskStorageMetric` and the value will get replaced in the map. Also technically that case should never actually happen - consider adding a warning log

Then this bit can just be:
```
final TaskStorageMetric taskMetric;
if (!metricsSeen.get(queryId).contains(taskId)) {
    taskMetric = new TaskStorageMetric(...);
    metricsSeen.get(queryId).add(taskId, new taskMetric);
    metricRegistry.addMetric(...)
} else {
    taskmetric = metricsSeen.get(queryId).get(taskId);
}
taskMetric.add(metric);
```

This should simplify removal too since you don't have to iterate the whole list.

Reply:
What case should never happen, calling add on an existing `TaskStorageMetric`? The rest of this makes sense and makes it cleaner